presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/TestRetryingPrestoThriftServiceProvider.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.connector.thrift.clientproviders;
+
+import com.facebook.presto.connector.thrift.ThriftConnectorConfig;
+import com.facebook.presto.connector.thrift.api.PrestoThriftService;
+import com.google.common.util.concurrent.ListeningScheduledExecutorService;
+import io.airlift.units.Duration;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import static com.google.common.util.concurrent.MoreExecutors.listeningDecorator;
+import static io.airlift.concurrent.Threads.threadsNamed;
+import static java.util.concurrent.Executors.newScheduledThreadPool;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static org.testng.Assert.assertEquals;
+
+public class TestRetryingPrestoThriftServiceProvider
+{
+    private ListeningScheduledExecutorService retryExecutor;
+
+    @BeforeClass
+    void init()
+    {
+        retryExecutor = listeningDecorator(newScheduledThreadPool(2, threadsNamed("test-thrift-retry-driver-%s")));
+    }
+
+    @AfterClass
+    void shutdown()
+    {
+        try {
+            retryExecutor.shutdownNow();
+        }
+        catch (Exception ignore) {
+        }
+    }
+
+    @Test
+    void testRetryService()
+            throws Exception
+    {
+        MockConnectedThriftServiceProvider mockServiceProvider = new MockConnectedThriftServiceProvider();
+        ThriftConnectorConfig config = new ThriftConnectorConfig()
+                .setMaxRetryAttempts(3)
+                .setMinRetrySleepTime(new Duration(10, MILLISECONDS))
+                .setMaxRetrySleepTime(new Duration(20, MILLISECONDS));
+
+        RetryingPrestoThriftServiceProvider retryingServiceProvider =
+                new RetryingPrestoThriftServiceProvider(mockServiceProvider, retryExecutor, config);
+
+        mockServiceProvider.setPendingFailuresToGenerate(2);
+
+        retryingServiceProvider.runOnAnyHost(PrestoThriftService::listSchemaNames);
+
+        assertEquals(mockServiceProvider.getOpenServiceCount(), 0);
```

Comment:
It doesn't seem to be verifying the case with removed `null` that you had before. Did it fail in that case?
If that's what you testing it might be good to count the number of times client was requested explicitly.

Reply:
It did fail in that case. I added check to explicitly check the number of open calls and close calls.